File: Medium/2391. Minimum Amount of Time to Collect Garbage/2391. Minimum Amount of Time to Collect Garbage.java
class Solution {
    public int garbageCollection(String[] garbage, int[] travel) {
        int acc = 0;
        // Calculate cumulative travel times
        int[] travelAccumulated = new int[travel.length + 1];
        for (int i = 0; i < travel.length; i++) {
            acc += travel[i];
            travelAccumulated[i + 1] = acc;
        }
        
        // Use nested loops to iterate through the garbage array and accumulate counts and last indices
        int p = 0, m = 0, g = 0, lastP = 0, lastM = 0, lastG = 0;
        for (int i = 0; i < garbage.length; i++) {
            String s = garbage[i];
            for (char c : s.toCharArray()) {
                switch (c) {
                    case 'P':
                        p += 1;
                        lastP = i;
                        break;
                    case 'M':
                        m += 1;
                        lastM = i;
                        break;
                    case 'G':
                        g += 1;
                        lastG = i;
                        break;
                }
            }
        }
        
        // Calculate total time
        return p + m + g + travelAccumulated[lastP] + travelAccumulated[lastM] + travelAccumulated[lastG];
    }
}

File: Medium/2391. Minimum Amount of Time to Collect Garbage/2391. Minimum Amount of Time to Collect Garbage.rs
impl Solution {
    pub fn garbage_collection(garbage: Vec<String>, travel: Vec<i32>) -> i32 {
        let mut acc = 0;
        // Calculate cumulative travel times
        let travel: Vec<i32> = std::iter::once(0).chain(travel.into_iter().map(|t| { acc += t; acc })).collect();
        
        // Use fold to iterate through the garbage vector and accumulate counts and last indices
        let (p, m, g, last_p, last_m, last_g) = garbage.into_iter().enumerate().fold(
            (0, 0, 0, 0, 0, 0),
            |(mut p, mut m, mut g, mut last_p, mut last_m, mut last_g), (i, s)| {
                for &b in s.as_bytes() {
                    match b {
                        b'P' => { p += 1; last_p = i; },
                        b'M' => { m += 1; last_m = i; },
                        _ => { g += 1; last_g = i; },
                    }
                }
                (p, m, g, last_p, last_m, last_g)
            }
        );
        
        // Calculate total time
        p + m + g + travel[last_p] + travel[last_m] + travel[last_g]
    }
}
